File: nvmolkit/substruct_library.py
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""GPU-resident molecule library for repeated substructure queries."""

from __future__ import annotations

from collections.abc import Iterable

from rdkit.Chem import Mol

from nvmolkit._substructLibrary import SubstructLibrary as _NativeSubstructLibrary
from nvmolkit.substructure import SubstructSearchConfig

__all__ = ["SubstructLibrary"]


class SubstructLibrary:
    """A collection of molecules prepared for repeated GPU substructure searches.

    Molecules added after construction become searchable after :meth:`finalize`.
    Queries continue to see the last successfully finalized collection while
    additional molecules are pending.
    """

    def __init__(
        self,
        chunkSize: int = 65_536,
        config: SubstructSearchConfig | None = None,
    ) -> None:
        """Create an empty library with the requested chunk size and search configuration."""
        if config is None:
            config = SubstructSearchConfig()
        self._native = _NativeSubstructLibrary(int(chunkSize), config._as_native())

    def __len__(self) -> int:
        """Return the number of molecules in the finalized collection."""
        return len(self._native)

    @property
    def pendingSize(self) -> int:
        """Number of added molecules awaiting finalization."""
        return self._native.pendingSize

    def addMol(self, molecule: Mol) -> int:
        """Add a molecule and return its stable library index."""
        return self._native.addMol(molecule)

    def addMols(self, molecules: Iterable[Mol]) -> list[int]:
        """Add molecules and return their stable library indices."""
        return self._native.addMols(molecules)

    def finalize(self) -> None:
        """Upload pending molecules and publish them for subsequent queries."""
        self._native.finalize()

    def getMatches(self, query: Mol, maxResults: int = -1) -> list[int]:
        """Return matching molecule indices in insertion order."""
        return self._native.getMatches(query, int(maxResults))

    def countMatches(self, query: Mol) -> int:
        """Return the number of molecules that match the query."""
        return self._native.countMatches(query)

    def hasMatch(self, query: Mol) -> bool:
        """Return whether any finalized molecule matches the query."""
        return self._native.hasMatch(query)

File: nvmolkit/substructLibrary.cpp
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

#include <GraphMol/ROMol.h>

#include <boost/noncopyable.hpp>
#include <boost/python.hpp>
#include <boost/python/stl_iterator.hpp>
#include <cstddef>
#include <vector>

#include "src/substruct/substruct_library.h"

namespace {

using namespace boost::python;

class ScopedGilRelease {
 public:
  ScopedGilRelease() : state_(PyEval_SaveThread()) {}
  ScopedGilRelease(const ScopedGilRelease&)            = delete;
  ScopedGilRelease& operator=(const ScopedGilRelease&) = delete;
  ~ScopedGilRelease() { PyEval_RestoreThread(state_); }

 private:
  PyThreadState* state_;
};

unsigned int addMolecule(nvMolKit::SubstructLibrary& library, const RDKit::ROMol& molecule) {
  const ScopedGilRelease release;
  return library.addMol(molecule);
}

list addMolecules(nvMolKit::SubstructLibrary& library, const object& molecules) {
  std::vector<object>              owners;
  std::vector<const RDKit::ROMol*> pointers;
  stl_input_iterator<object>       iterator(molecules), end;
  for (; iterator != end; ++iterator) {
    owners.push_back(*iterator);
    pointers.push_back(extract<const RDKit::ROMol*>(owners.back()));
  }

  std::vector<unsigned int> ids;
  {
    const ScopedGilRelease release;
    ids = library.addMols(pointers);
  }

  list result;
  for (const unsigned int id : ids) {
    result.append(id);
  }
  return result;
}

void finalize(nvMolKit::SubstructLibrary& library) {
  const ScopedGilRelease release;
  library.finalize();
}

list getMatches(nvMolKit::SubstructLibrary& library, const RDKit::ROMol& query, int maxResults) {
  std::vector<unsigned int> ids;
  {
    const ScopedGilRelease release;
    ids = library.getMatches(query, maxResults);
  }
  list result;
  for (const unsigned int id : ids) {
    result.append(id);
  }
  return result;
}

std::size_t countMatches(nvMolKit::SubstructLibrary& library, const RDKit::ROMol& query) {
  const ScopedGilRelease release;
  return library.countMatches(query);
}

bool hasMatch(nvMolKit::SubstructLibrary& library, const RDKit::ROMol& query) {
  const ScopedGilRelease release;
  return library.hasMatch(query);
}

}  // namespace

BOOST_PYTHON_MODULE(_substructLibrary) {
  import("rdkit.Chem.rdchem");
  import("nvmolkit._substructure");

  class_<nvMolKit::SubstructLibrary, boost::noncopyable>(
    "SubstructLibrary",
    init<std::size_t, nvMolKit::SubstructSearchConfig>(
      (arg("chunkSize") = 65536, arg("config") = nvMolKit::SubstructSearchConfig())))
    .def("addMol", &addMolecule)
    .def("addMols", &addMolecules)
    .def("finalize", &finalize)
    .def("getMatches", &getMatches, (arg("query"), arg("maxResults") = -1))
    .def("countMatches", &countMatches)
    .def("hasMatch", &hasMatch)
    .def("__len__", &nvMolKit::SubstructLibrary::size)
    .add_property("pendingSize", &nvMolKit::SubstructLibrary::pendingSize);
}
